Mathlib/Analysis/PSeries.lean

```diff
@@ -17,132 +17,218 @@ converges if and only if so does `∑ k, 2 ^ k f (2 ^ k)`. We prove this test in
-    simpa [h, ENNReal.add_eq_top, ENNReal.mul_eq_top] using ENNReal.tsum_condensed_le hf
+    obtain ⟨C, hCpos, hC⟩ := ENNReal.tsum_schlomilch_le hf h_pos strict_mono h_succ_diff
+    have C_nonzero : C ≠ 0 := ne_of_gt hCpos
+    have : (↑(u 1) - ↑(u 0)) * ↑(f (u 0)) + ↑C * ∑' (k : ℕ), ↑(f k) = ∞ := by exact eq_top_mono hC h
+    simpa [ENNReal.add_eq_top, ENNReal.mul_ne_top, ENNReal.mul_eq_top, C_nonzero]
   · replace hf : ∀ m n, 0 < m → m ≤ n → (f n : ℝ≥0∞) ≤ f m := fun m n hm hmn =>
       ENNReal.coe_le_coe.2 (hf hm hmn)
-    simpa [h, ENNReal.add_eq_top] using ENNReal.le_tsum_condensed hf
-#align nnreal.summable_condensed_iff NNReal.summable_condensed_iff
-
+    have : ∑ k in range (u 0), ↑(f k) ≠ ∞ := by
+      apply ne_top_of_lt (ENNReal.sum_lt_top fun a _ => coe_ne_top)
+    simpa [h, ENNReal.add_eq_top, this] using ENNReal.le_tsum_schlomilch hf h_pos strict_mono
 end NNReal
 
-/-- Cauchy condensation test for series of nonnegative real numbers. -/
+/-- for series of nonnegative real numbers. -/
+theorem summable_schlomilch_iff_of_nonneg {u : ℕ → ℕ} {f : ℕ → ℝ} (h_nonneg : ∀ n, 0 ≤ f n)
+    (hf : ∀ ⦃m n⦄, 0 < m → m ≤ n → f n ≤ f m) (h_pos : ∀ n, 0 < u n)
+    (strict_mono : ∀ ⦃m n⦄, 0 ≤ m → m < n → u m < u n) (h_succ_diff : succ_diff_bounded u) :
+    (Summable fun k : ℕ => (u (k + 1) - (u k : ℝ)) * f (u k)) ↔ Summable f := by
+  lift f to ℕ → ℝ≥0 using h_nonneg
+  simp only [NNReal.coe_le_coe] at *
+  have : ∀ k, (u (k + 1) - (u k : ℝ)) = ((u (k + 1) : ℝ≥0) - (u k : ℝ≥0) : ℝ≥0) := by
+    intro k
+    rw [@coe_sub_def]
+    rw [max_eq_left]
+    simp only [NNReal.coe_nat_cast]
+    refine sub_nonneg_of_le ?_
+    norm_cast
+    exact (strict_mono k.zero_le k.lt_succ_self).le
+  simp_rw [this]
+  exact_mod_cast NNReal.summable_schlomilch_iff hf h_pos strict_mono h_succ_diff
+
 theorem summable_condensed_iff_of_nonneg {f : ℕ → ℝ} (h_nonneg : ∀ n, 0 ≤ f n)
     (h_mono : ∀ ⦃m n⦄, 0 < m → m ≤ n → f n ≤ f m) :
     (Summable fun k : ℕ => (2 : ℝ) ^ k * f (2 ^ k)) ↔ Summable f := by
-  lift f to ℕ → ℝ≥0 using h_nonneg
-  simp only [NNReal.coe_le_coe] at *
-  exact_mod_cast NNReal.summable_condensed_iff h_mono
-#align summable_condensed_iff_of_nonneg summable_condensed_iff_of_nonneg
+  have h_pos : ∀ (n : ℕ), 0 < 2 ^ n := fun n => pow_pos zero_lt_two n
+  have strict_mono : ∀ ⦃m n⦄, 0 ≤ m → m < n → 2 ^ m < 2 ^ n := fun m n _ hmn =>
+    pow_lt_pow ?h hmn
+  have h_succ_diff : succ_diff_bounded (fun n => 2 ^ n) := by
+    use 2
+    norm_num
+    intro n
+    simp [pow_succ, two_mul]
+  · have : ∀ k, (2 : ℝ) ^ k * f (2 ^ k) = (2 ^ (k + 1) - 2 ^ k) * f (2 ^ k) := by
+      intro k
+      refine mul_eq_mul_right_iff.mpr ?_
+      left
+      have foo : 2 ^ k ≤ 2 ^ (k + 1) := by
+        exact le_of_lt (pow_lt_pow one_lt_two (Nat.lt_succ_self k))
+      norm_cast
+      simp [pow_succ, two_mul]
+    simp_rw [this]
+    exact_mod_cast summable_schlomilch_iff_of_nonneg h_nonneg h_mono h_pos strict_mono h_succ_diff
+  exact one_lt_two
```

Comment:
reasoning in reverse is sometimes quite helpful. Maybe you haven't learned some of these tactics yet:
```suggestion
    pow_lt_pow (Nat.lt_succ_self 1) hmn
  have h_succ_diff : succ_diff_bounded (fun n => 2 ^ n) :=
    ⟨2, by norm_num, by norm_num, by simp [pow_succ, two_mul]⟩
  convert summable_schlomilch_iff_of_nonneg h_nonneg h_mono h_pos strict_mono h_succ_diff
  simp only [Nat.cast_pow, Nat.cast_ofNat]
  simp [pow_succ, two_mul]
```